content/security/authentication.md

```diff
@@ -984,3 +984,96 @@ whoAmI(@CurrentUser() user: User) {
+}
+```
+> info **Hint** You *must* include the `handleRequest()` method, and return the user object. Otherwise, Passport will not store the user in the session, and you will have an error similar to this:
+>
+> `Error: Failed to find request token in session`
+
+This is an example Controller class. When the user accesses the `/api/auth/twitter` endpoint, they will be automatically redirected by the `TwitterAuthGuard` to the Twitter OAuth webpage. Twitter then sends the user back to the `/api/auth/twitter/callback` endpoint, and as shown below, the user is then authenticated, and the user object is now in the `Request` object.
+```typescript
+@Controller('api/auth')
+export class AuthController {
+  constructor() {}
+
+  @Get('twitter')
+  @UseGuards(TwitterAuthGuard)
+  async twitter() {
+    throw new UnauthorizedException(); // Guard redirects, this will never be hit
+  }
+
+  @Get('twitter/callback')
+  @UseGuards(TwitterAuthGuard)
+  async twitterCallback(@Req() req: Request, @Res() res: Response) {
+    res.send({user: req.user}); // req.user will now contain the user from the 
+  }
+}
+```
+
+Here is an example `PassportStrategy` implementation for Twitter. You must pass the correct passport options into the constructor's `super` call.
+```typescript
+@Injectable()
+export class TwitterStrategy extends PassportStrategy(Strategy) {
+  constructor(private configService: ConfigService) {
+    super({
+      consumerKey: configService.get<string>('TWITTER_API_KEY'),
+      consumerSecret: configService.get<string>('TWITTER_API_SECRET'),
+      callbackURL: 'http://127.0.0.1:3000/api/auth/twitter/callback',
+      passReqToCallback: true,
+      // includeEmail: true,
+    });
+  }
+
+  validate(
+    req: Request,
+    accessToken: string,
+    refreshToken: string,
+    profile: Profile,
+    done,
+  ) {
+    const twitterUser = {
+      id: profile.id,
+      username: profile.username,
+      displayName: profile.displayName,
+      email: profile.emails?.shift().value,
+      photo: profile.photos?.shift().value,
+    };
+
+    // This is where you would store the Twitter user
+    // User.findOrCreate({ twitterId: profile.id }, function (err, user) {
+    //   return done(err, user);
+    // });
+    
+    done(null, twitterUser); // Must call the callback with the user object
+  }
+}
+```
+
+It is important to note that you may only gather the user's email if you have special permissions from Twitter. If you are granted permissions, you may use the `includeEmail: true,` property.
+> info **Hint** You will need to change the `callbackURL` property to your host domain if not running locally.
+>
+> It is also important to recognize that you must use the address of `http://127.0.0.1` instead of local host, when running locally. These values must match what you provide on the twitter dev page.
+>
+> You may obtain your Twitter API Key and Secret Key from the twitter developer portal [here](https://developer.twitter.com/apply-for-access).
```

Comment:
For sure, I think each provider sends back different information though, and has different options. Do you think I should just use your wording, and remove the code examples?

Reply:
@jmcdo29 So I'm reading the twitter documentation...they don't even support user authentication via OAuth2.

"Please note that OAuth 1.0a is still required to issues requests on behalf of users"

"Application-only authentication doesn't include any user-context and is a form of authentication where an application makes API requests on its own behalf. This method is for developers that just need read-only access to public information."

So, I don't really think that writing a general "oauth2" example in the docs would really make sense for most people, and could be even more confusing, especially since the documentation is under the heading of "Authentication", e.g. user authentication.

It appears that different corporations use different authentication strategies for each, e.g. twitter uses oauth1 for user authentication, and facebook might use oauth2 for user authentication

Reply:
> So I don't think that writing a general oauth2 example... would really make sense for most people

Hmm, possibly. I'm hesitant about putting a specific example for Twitter though, as we'll start getting requests for Facebook, Google, LinkedIn, Discord, and who knows what else. Would it be possible to just mention the use of the `handleRequest` and _why_ it's necessary to return the `user` instead of the default functionality Nest has right there? 

> Passport also makes it pretty clear...

Man, I wish people would look at the underlying package's documentation. 90% of our issues seem to come from people not reading the underlying docs first. But you're right, they do mention that each provider is different.

Reply:
Sure, I'll close this.